File: src/service/compute/utils/helpers_test.go
/*
SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
*/

package utils_test

import (
	"context"
	"errors"
	"io"
	"testing"

	"go.corp.nvidia.com/osmo/service/compute/utils"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
)

func TestExtractBackendName(t *testing.T) {
	tests := []struct {
		name        string
		ctx         context.Context
		expected    string
		shouldError bool
	}{
		{
			name:        "valid backend name",
			ctx:         metadata.NewIncomingContext(context.Background(), metadata.Pairs("backend-name", "my-backend")),
			expected:    "my-backend",
			shouldError: false,
		},
		{
			name:        "missing metadata",
			ctx:         context.Background(),
			expected:    "",
			shouldError: true,
		},
		{
			name:        "empty backend name",
			ctx:         metadata.NewIncomingContext(context.Background(), metadata.Pairs("backend-name", "")),
			expected:    "",
			shouldError: true,
		},
		{
			name:        "backend-name not in metadata",
			ctx:         metadata.NewIncomingContext(context.Background(), metadata.Pairs("other-key", "value")),
			expected:    "",
			shouldError: true,
		},
		{
			name:        "multiple backend names (first is used)",
			ctx:         metadata.NewIncomingContext(context.Background(), metadata.Pairs("backend-name", "first", "backend-name", "second")),
			expected:    "first",
			shouldError: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := utils.ExtractBackendName(tt.ctx)

			if tt.shouldError {
				if err == nil {
					t.Errorf("expected error but got nil")
				}
			} else {
				if err != nil {
					t.Errorf("unexpected error: %v", err)
				}
				if result != tt.expected {
					t.Errorf("expected %q, got %q", tt.expected, result)
				}
			}
		})
	}
}

func TestIsExpectedClose(t *testing.T) {
	tests := []struct {
		name     string
		err      error
		expected bool
	}{
		{
			name:     "nil error",
			err:      nil,
			expected: false,
		},
		{
			name:     "io.EOF",
			err:      io.EOF,
			expected: true,
		},
		{
			name:     "context.Canceled",
			err:      context.Canceled,
			expected: true,
		},
		{
			name:     "status codes.Canceled",
			err:      status.Error(codes.Canceled, "canceled"),
			expected: true,
		},
		{
			name:     "other error",
			err:      errors.New("some error"),
			expected: false,
		},
		{
			name:     "status codes.Internal",
			err:      status.Error(codes.Internal, "internal error"),
			expected: false,
		},
		{
			name:     "status codes.Unknown",
			err:      status.Error(codes.Unknown, "unknown error"),
			expected: false,
		},
		{
			name:     "status codes.InvalidArgument",
			err:      status.Error(codes.InvalidArgument, "invalid argument"),
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.IsExpectedClose(tt.err)
			if result != tt.expected {
				t.Errorf("expected %v, got %v", tt.expected, result)
			}
		})
	}
}

File: src/service/agent/helpers_test.py
"""
SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
"""

import unittest
from unittest import mock

from src.service.agent import helpers
from src.utils import backend_messages
from src.utils.job import task


class OTGStatusUpdateTest(unittest.TestCase):

    def test_task_updates_are_forwarded_to_pod_status_path(self):
        message = backend_messages.OTGStatusBody(
            workflow_uuid='workflow-uuid',
            group_name='group-a',
            namespace='osmo-vpan',
            name='otg-a',
            phase='Running',
            task_status_updates=[
                backend_messages.OTGTaskStatusUpdateBody(
                    workflow_uuid='workflow-uuid',
                    task_uuid='task-uuid',
                    retry_id=2,
                    container='user',
                    status='RUNNING',
                    backend='backend-a',
                )
            ],
        )
        forwarded = []

        with mock.patch.object(
                helpers, 'queue_update_group_job', side_effect=lambda _postgres, body:
                forwarded.append(body)):
            helpers.queue_otg_status_update_job(mock.Mock(), message)

        self.assertEqual(len(forwarded), 1)
        self.assertEqual(forwarded[0].workflow_uuid, 'workflow-uuid')
        self.assertEqual(forwarded[0].task_uuid, 'task-uuid')
        self.assertEqual(forwarded[0].status, 'RUNNING')

    def test_group_fallback_queues_update_group_when_no_task_updates_exist(self):
        postgres = mock.Mock()
        postgres.execute_fetch_command.return_value = [{
            'workflow_id': 'workflow-a',
            'workflow_uuid': 'workflow-uuid',
            'submitted_by': 'user-a',
        }]
        message = backend_messages.OTGStatusBody(
            workflow_uuid='workflow-uuid',
            group_name='group-a',
            namespace='osmo-vpan',
            name='otg-a',
            phase='Failed',
            message='controller failed',
        )

        with mock.patch.object(helpers.jobs, 'UpdateGroup') as update_group_class:
            helpers.queue_otg_status_update_job(postgres, message)

        update_group_class.assert_called_once()
        update_group_class.return_value.send_job_to_queue.assert_called_once()
        self.assertEqual(
            update_group_class.call_args.kwargs['status'],
            task.TaskGroupStatus.FAILED_SERVER_ERROR,
        )
        self.assertEqual(update_group_class.call_args.kwargs['message'], 'controller failed')


if __name__ == '__main__':
    unittest.main()
